### ISSUE TITLE
Add option to JsonSerializer to use Spring primary ObjectMapper bean 

### ISSUE THREAD

**Issue by yoransabern** [spring-projects/spring-kafka#680]
Can we get an option to have the producer value-serializer use the ObjectMapper bean instantiated by Spring boot instead of having to configure a custom JsonSerializer and ProducerFactory bean?

This is usefull to do all kafka configuration through the properties instead of having to do custom code configuration, and I think most spring libraries already get the primary ObjectMapper bean if it is available.

**Comment by yoransabern** [comment]
Same goes for the JsonDeserializer.

**Comment by yoransabern** [comment]
Some explanation behind this requirement:

To properly serialize and deserialize the Java.time.Instant objects, the objectmapper requires a java time module to be set.

It would also be usefull to specify modules to be set for the Object mappers used in the serializers.

**Comment by garyrussell** [comment]
That can't be done by this project, it would need to be done, somehow, in Boot's auto configuration.

When providing the (De)Serializer class through Kafka properties, the Kafka client instantiates the object, not Spring. Kafka knows nothing about Spring or Boot.

The only way for Spring to do the instantiation is through a custom Producer/Consumer factory, taking the (De)Serializers as constructor arguments.

So, I suggest moving this request to spring-boot; perhaps some new properties could trigger conditional injection of key/value (de)serializers.

**Comment by yoransabern** [comment]
Ah alright, I guess it would not be an option to set the Java Time module on the objectmapper which is created in the serializer based on a spring.kafka property?

**Comment by artembilan** [comment]
> I guess it would not be an option to set the Java Time module on the objectmapper which is created in the serializer based on a spring.kafka property?

Well, I don't think so. 
We really can do in the `JsonSerializer` ctor - something similar what we have in Spring Integration with its `Jackson2JsonObjectMapper`: https://github.com/spring-projects/spring-integration/blob/master/spring-integration-core/src/main/java/org/springframework/integration/support/json/Jackson2JsonObjectMapper.java#L154

**Comment by mle-enso** [comment]
Is there any news on this one?

**Comment by artembilan** [comment]
I think the solution I provide in the #1053 is a good compromise where we don't have anything auto-configured, but really would like to rely on the classpath.
The story about auto-configuration JSON (de)serializers in Spring Boot is fully different issue and belongs that project.

Thanks

**Comment by jbotuck** [comment]
> When providing the (De)Serializer class through Kafka properties, the Kafka client instantiates the object, not Spring. Kafka knows nothing about Spring or Boot.
> 
> The only way for Spring to do the instantiation is through a custom Producer/Consumer factory, taking the (De)Serializers as constructor arguments.

Why can't we have the constructor of the(De)Serializer class look for an ObjectMapper bean in the application context before constructing its own?

**Comment by artembilan** [comment]
Why existing constructor to inject is not enough:
```
	/**
	 * Construct an instance with the provided {@link ObjectMapper}.
	 * @param objectMapper a custom object mapper.
	 */
	public JsonDeserializer(ObjectMapper objectMapper)
```
?

You can use this in your application like this:
```
@Bean 
Deserializer jsonDeserializer(ObjectMapper objectMapper) {
    return new JsonDeserializer(objectMapper);
}
```

**Comment by jbotuck** [comment]
Because now I have to construct the producerFactory in code instead of just  setting `spring.kafka.producer.value-serializer=org.springframework.kafka.support.serializer.JsonSerializer`

**Comment by garyrussell** [comment]
>just setting `spring.kafka.producer.value-serializer=org.springframework.kafka.support.serializer.JsonSerializer`

When you use a kafka consumer property like that, Kafka instantiates the deserializer, not Spring. Kafka knows nothing about the application context the clients are running in.

**Comment by Krandelbord** [comment]
In which version it was fixed ?

**Comment by artembilan** [comment]
Not sure what the question. The related issue was closed in version `2.3`: https://github.com/spring-projects/spring-kafka/issues/680
